### PR TITLE
docs: explain how to implement the four client state functions which allow for regular updates and misbehaviour handling

### PR DESCRIPTION
**Pull request by charleenfei**
## Description

closes: #1854 

### Commit Message / Changelog Entry

```bash
docs: explain how to implement the four client state functions which allow for regular updates and misbehaviour handling
```

see the [guidelines](https://github.com/cosmos/ibc-go/blob/main/CONTRIBUTING.md#commit-messages) for commit messages. (view raw markdown for examples)

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#pr-targeting)).
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules/10-structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#testing).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/CONTRIBUTING.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

File: docs/ibc/light-clients/updates-and-misbehaviour.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
+<!--
+order: 5
+-->
+
+# Handling Updates and Misbehaviour
+
+The functions for handling updates to a light clients and evidence of misbehaviour are all found in the [`ClientState`](https://github.com/cosmos/ibc-go/blob/v6.0.0/modules/core/exported/client.go#L40) interface, and will be discussed below.
+
+It is important to note that `Misbehaviour` in this particular context is referring to misbehaviour on the chain level intended to fool the light client. This will be defined by each light client. For example, the Tendermint light client [defines `Misbehaviour`](https://github.com/cosmos/ibc-go/blob/main/modules/light-clients/07-tendermint/misbehaviour.go) as a situation where two conflicting `Headers` have been submitted to update a client's `ConsensusState` within the same trusting period.
```
Comment (colin-axner) [inline docs/ibc/light-clients/updates-and-misbehaviour.md:9]: We could note BFT time misbehaviour here as well

File: docs/ibc/light-clients/updates-and-misbehaviour.md [hunk 2]
```diff
@@ -0,0 +1,38 @@
+<!--
+order: 5
+-->
+
+# Handling Updates and Misbehaviour
```
Comment (colin-axner) [inline docs/ibc/light-clients/updates-and-misbehaviour.md:5]: could we add pseudocode from the `UpdateClient` function which shows the usage of these 4 functions?

Something like:
```go
    if err := clientState.VerifyClientMessage(clientMessage); err != nil {
        return err
    }
    
    foundMisbehaviour := clientState.CheckForMisbehaviour(clientMessage)
    if foundMisbehaviour {
        clientState.UpdateStateOnMisbehaviour(clientMessage)
        // emit misbehaviour event
        return 
    }
    
    clientState.UpdateState(clientMessage) // expects no-op on duplicate header
    // emit update event
    return
}
```

**Review by colin-axner — COMMENTED** [review]
Ay nice work! Left some suggestions. It's great to have documentation on these functions

File: docs/ibc/light-clients/updates-and-misbehaviour.md [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+## UpdateStateOnMisbehaviour 
+
+`UpdateStateOnMisbehaviour` should perform appropriate state changes on a client state given that misbehaviour has been detected and verified. This method should only be called when misbehaviour is detected, as it does not perform any misbehaviour checks. Notably, it should freeze the client so that calling the `Status()` function on the associated client state no longer returns `Active`.
+
+For an example of a `UpdateStateOnMisbehaviour` implementation, please check the [Tendermint light client](https://github.com/cosmos/ibc-go/blob/main/modules/light-clients/07-tendermint/update.go#L197)
+
+## UpdateState
```
Comment (crodriguezvega) [inline docs/ibc/light-clients/updates-and-misbehaviour.md:33]:
```suggestion
## `UpdateState`
```

**Review by crodriguezvega — COMMENTED** [review]
Thank you, @charleenfei. Left just some nits.

File: docs/ibc/light-clients/updates-and-misbehaviour.md [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+## `UpdateState`
+
+`UpdateState` updates and stores as necessary any associated information for an IBC client, such as the `ClientState` and corresponding `ConsensusState`. It should perform a no-op on duplicate updates.
+
+It assumes the `ClientMessage` has already been verified.
+
+For an example of a `UpdateState` implementation, please check the [Tendermint light client](https://github.com/cosmos/ibc-go/blob/main/modules/light-clients/07-tendermint/update.go#L131).
+
+## Reference Example
```
Comment (tmsdkeys) [inline docs/ibc/light-clients/updates-and-misbehaviour.md:41]: Not sure if I'd go with 'reference example' as title here. Throughout the page, we already use the `07-tendermint` as reference example for the light client implementation so using 'reference example' here again when looking at `02-client` feels a little confusing.

How about "Putting it all together" or something similar as section title containing the pseudo code like @colin-axner mentioned above and then also refer to `02-client`?

File: docs/ibc/light-clients/updates-and-misbehaviour.md [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+## `CheckForMisbehaviour`
+
+Checks for evidence of a misbehaviour in `Header` or `Misbehaviour` type. It assumes the `ClientMessage` has already been verified.
+
+For an example of a `CheckForMisbehaviour` implementation, please check the [Tendermint light client](https://github.com/cosmos/ibc-go/blob/main/modules/light-clients/07-tendermint/misbehaviour_handle.go#L18).
+
+For example, the Tendermint light client [defines `Misbehaviour`](https://github.com/cosmos/ibc-go/blob/main/modules/light-clients/07-tendermint/misbehaviour.go) as two different types of situations: a situation where two conflicting `Headers` with the same height have been submitted to update a client's `ConsensusState` within the same trusting period, or that the two conflicting `Headers` have been submitted at different heights but the consensus states are not in the correct monotonic time ordering (BFT time violation). More explicitly, updating to a new height must have a timestamp greater than the previous consensus state, or, if inserting a consensus at a past height, then time must be less than those heights which come after and greater than heights which come before.
```
Comment (tmsdkeys) [inline docs/ibc/light-clients/updates-and-misbehaviour.md:25]:
```suggestion
> For example, the Tendermint light client [defines `Misbehaviour`](https://github.com/cosmos/ibc-go/blob/main/modules/light-clients/07-tendermint/misbehaviour.go) as two different types of situations: a situation where two conflicting `Headers` with the same height have been submitted to update a client's `ConsensusState` within the same trusting period, or that the two conflicting `Headers` have been submitted at different heights but the consensus states are not in the correct monotonic time ordering (BFT time violation). More explicitly, updating to a new height must have a timestamp greater than the previous consensus state, or, if inserting a consensus at a past height, then time must be less than those heights which come after and greater than heights which come before.
```
Maybe some extra formating to make clear this is a note?

**Review by tmsdkeys — APPROVED** [review]
Nice work @charleenfei ! Left a few minor suggestions to consider

**Review by AdityaSripal — APPROVED** [review]

**Pull request merged**